src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -58,57 +58,169 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
+                // csharp and razor ranges have the same span; skip the C# item
+                ++indexCsharp;
+            }
+            else if (comparison < 0)
+            {
+                newList.Add(currentRazorRange);
+                ++indexRazor;
+            }
+            else
+            {
+                newList.Add(currentCsharpRange);
+                ++indexCsharp;
+            }
+        }
+
+        while (indexRazor < razorRanges.Count)
+        {
+            newList.Add(razorRanges[indexRazor]);
+            ++indexRazor;
+        }
 
-        return newList;
+        while (indexCsharp < csharpRanges.Count)
+        {
+            newList.Add(csharpRanges[indexCsharp]);
+            ++indexCsharp;
+        }
+
+#if DEBUG
+        // Verify that the result of merging above matches the old algorithm of doing a 'full sort'.        
+        using var _2 = ListPool<SemanticRange>.GetPooledObject(out var fullSortList);
+
+        fullSortList.AddRange(razorRanges);
+        fullSortList.AddRange(csharpRanges);
+
+        fullSortList.Sort((left, right) =>
+        {
+            var rangeCompare = left.CompareTo(right);
+            if (rangeCompare != 0)
+            {
+                return rangeCompare;
+            }
+
+            // If we have ranges that are the same, we want a Razor produced token to win over a non-Razor produced token
+            if (left.FromRazor && !right.FromRazor)
+            {
+                return -1;
+            }
+            else if (right.FromRazor && !left.FromRazor)
+            {
+                return 1;
+            }
+
+            return 0;
+        });
+
+        for (var idx = 1; idx < fullSortList.Count; ++idx)
+        {
+            if (fullSortList[idx].CompareTo(fullSortList[idx - 1]) == 0)
+            {
+                fullSortList.RemoveAt(idx);
+                idx--;
+            }
+        }
+
+        Debug.Assert(fullSortList.Count == newList.Count,$"After sort and removing duplicates (favoring Razor over C#), the lists should be equal size.  Fast algorithm: ${newList.Count} Full sort: ${fullSortList.Count}");
+
+        for (var idx = 0; idx < fullSortList.Count; ++idx)
+        {
+            Debug.Assert(Equals(newList[idx], fullSortList[idx]), $"Difference between the full sort and the merge of sorted lists at index ${idx}");
+        }
+#endif
+
+        return pooledList;
     }
 
-    // Internal and virtual for testing only
-    internal virtual async Task<List<SemanticRange>?> GetCSharpSemanticRangesAsync(
+    /// <summary>
+    /// Returns a sorted list of SemanticRange objects representing the semantic info
+    /// for the C# regions of the Razor file.
+    ///
+    /// Internal and virtual to enable testing; for running product, no need to override.
+    /// </summary>
+    /// <returns>List of Semantic range sorted by position</returns>
+    internal virtual async Task<PooledObject<List<SemanticRange>>?> GetCSharpSemanticRangesAsync(
```

Comment:
I completely agree and I would have blocked this PR if I'd seen this. If you have a scenario where a list needs to be passed outside of a method, a pooled list should never be passed back because the lifetime of that pooled list isn't known to the calling code.

Reply:
In general, it is probably better to determine whether `List<T>` is really the best type to be using. If the collection isn't going to modified outside of the collection, it might be better to just return an `ImmutableArray<T>`. And of course, if you really need a list to be a workhorse and use across several methods, just create a list.